File: DynamicProgramming/maxPathSum.js
class TreeNode {
    constructor(val) {
        this.val = val
        this.left = this.right = null
    }
}

function memoizedMaxPathSum(root) {
    let maxSum = Number.MIN_SAFE_INTEGER

    function dfs(node) {
        if (!node) return 0

        const leftSum = Math.max(dfs(node.left), 0)
        const rightSum = Math.max(dfs(node.right), 0)

        const currentSum = node.val + leftSum + rightSum

        maxSum = Math.max(maxSum, currentSum)

        return node.val + Math.max(leftSum, rightSum)
    }

    dfs(root)
    return maxSum
}

function nonMemoizedMathPathSum(root) {
    let maxSum = Number.MIN_SAFE_INTEGER;
  
    function findMaxPath(node) {
      if (!node) return 0;
  
      const leftSum = findMaxPath(node.left);
      const rightSum = findMaxPath(node.right);
  
      const currentSum = Math.max(node.val, node.val + leftSum, node.val + rightSum);
  
      maxSum = Math.max(maxSum, currentSum, node.val + leftSum + rightSum);
  
      return Math.max(node.val, node.val + leftSum, node.val + rightSum);
    }
  
    findMaxPath(root);
    return maxSum;
  }
  
const root = new TreeNode(10)
root.left = new TreeNode(2)
root.right = new TreeNode(10)
root.left.left = new TreeNode(20)
root.left.right = new TreeNode(1)
root.right.right = new TreeNode(-25)
root.right.right.left = new TreeNode(3)
root.right.right.right = new TreeNode(4)


console.time('nonMemoizedMathPathSum')
console.log(nonMemoizedMathPathSum(root))
console.timeEnd("nonMemoizedMathPathSum")

console.time('memoizedMaxPathSum')
console.log(memoizedMaxPathSum(root))
console.timeEnd("memoizedMaxPathSum")

File: DynamicProgramming/maxPathSum.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	maxSum := math.MinInt32
	findMaxPath(root, &maxSum)
	return maxSum
}

func findMaxPath(node *TreeNode, maxSum *int) int {
	if node == nil {
		return 0
	}

	leftSum := max(findMaxPath(node.Left, maxSum), 0)
	rightSum := max(findMaxPath(node.Right, maxSum), 0)

	currentSum := node.Val + leftSum + rightSum
	*maxSum = max(*maxSum, currentSum)

	return node.Val + max(leftSum, rightSum)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	// Example binary tree
	root := &TreeNode{Val: 10}
	root.Left = &TreeNode{Val: 2}
	root.Right = &TreeNode{Val: 10}
	root.Left.Left = &TreeNode{Val: 20}
	root.Left.Right = &TreeNode{Val: 1}
	root.Right.Right = &TreeNode{Val: -25}
	root.Right.Right.Left = &TreeNode{Val: 3}
	root.Right.Right.Right = &TreeNode{Val: 4}

	result := maxPathSum(root)
	fmt.Println(result) // Output: 42 (The maximum path sum is obtained by the path: 20 -> 2 -> 10 -> 10)
}
